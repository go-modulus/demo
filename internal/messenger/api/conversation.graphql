extend type Query {
    conversation(id: Uuid!): ConversationResult!
    myConversations(first: Int! = 25, after: String): MyConversationsResult!
}

extend type Mutation {
    createOneToOneConversation(receiverId: Uuid!): CreateOneToOneConversationResult!
}

union ConversationResult = ConversationBox | ErrUnknownConversation | ErrInvalidInput | ErrUnauthorized

type ErrUnknownConversation implements Error {
    message: String!
}

union MyConversationsResult = ConversationList | ErrInvalidInput | ErrUnauthorized

union CreateOneToOneConversationResult = OneToOneConversation | ErrUnknownReceiver | ErrInvalidInput | ErrUnauthorized

type ErrUnknownReceiver implements Error {
    message: String!
}

union Conversation = OneToOneConversation

type ConversationBox {
    value: Conversation!
}

type OneToOneConversation {
    id: Uuid!
    draft: Draft @goField(forceResolver: true)
    lastMessage: Message @goField(forceResolver: true)
    createdAt: MilliTimestamp!
}

type ConversationEdge {
    cursor: String!
    node: Conversation!
}

type ConversationList {
    edges: [ConversationEdge]!
    pageInfo: PageInfo!
}

directive @goField(
    forceResolver: Boolean
    name: String
) on INPUT_FIELD_DEFINITION | FIELD_DEFINITION | INTERFACE