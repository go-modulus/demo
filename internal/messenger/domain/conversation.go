package domain

import (
	"context"
	"database/sql"
	"fmt"
	"github.com/gofrs/uuid"
	"time"
)

type ErrConversationNotFound struct {
	ConversationId uuid.UUID
}

func (e ErrConversationNotFound) Error() string {
	return fmt.Sprintf("conversation with id %s not found", e.ConversationId)
}

type Conversation struct {
	Id         uuid.UUID
	SenderId   uuid.UUID
	ReceiverId uuid.UUID
	UpdatedAt  time.Time
	CreatedAt  time.Time
}

func (c Conversation) IsParticipant(personId uuid.UUID) bool {
	return c.SenderId == personId || c.ReceiverId == personId
}

type ConversationRepository interface {
	Get(ctx context.Context, id uuid.UUID) (*Conversation, error)
	Paginate(ctx context.Context, viewer uuid.UUID, first int, after sql.NullString) (*Page[*Conversation], error)

	GetOrCreate(ctx context.Context, sender uuid.UUID, receiver uuid.UUID) (*Conversation, error)
}

type ConversationCreator struct {
	conversations ConversationRepository
}

func NewConversationCreator(conversations ConversationRepository) *ConversationCreator {
	return &ConversationCreator{conversations: conversations}
}

func (c *ConversationCreator) Create(ctx context.Context, sender uuid.UUID, receiver uuid.UUID) (*Conversation, error) {
	conversation, err := c.conversations.GetOrCreate(ctx, sender, receiver)
	if err != nil {
		return nil, err
	}

	return conversation, nil
}

type ConversationViewer struct {
	conversations ConversationRepository
	policy        *ConversationPolicy
}

func NewConversationViewer(conversations ConversationRepository, policy *ConversationPolicy) *ConversationViewer {
	return &ConversationViewer{conversations: conversations, policy: policy}
}

func (c *ConversationViewer) View(ctx context.Context, viewer uuid.UUID, id uuid.UUID) (*Conversation, error) {
	conversation, err := c.conversations.Get(ctx, id)
	if err != nil {
		return nil, err
	}

	if !c.policy.CanView(viewer, conversation) {
		return nil, ErrConversationNotFound{ConversationId: id}
	}

	return conversation, nil
}

func (c *ConversationViewer) Paginate(
	ctx context.Context,
	viewer uuid.UUID,
	first int,
	after sql.NullString,
) (*Page[*Conversation], error) {
	page, err := c.conversations.Paginate(ctx, viewer, first, after)
	if err != nil {
		return nil, err
	}

	return page, nil
}

type ConversationPolicy struct{}

func NewConversationPolicy() *ConversationPolicy {
	return &ConversationPolicy{}
}

func (c *ConversationPolicy) CanView(personId uuid.UUID, conversation *Conversation) bool {
	if conversation == nil {
		return false
	}

	return conversation.IsParticipant(personId)
}

func (c *ConversationPolicy) CanSendMessages(personId uuid.UUID, conversation *Conversation) bool {
	if conversation == nil {
		return false
	}

	return conversation.IsParticipant(personId)
}
