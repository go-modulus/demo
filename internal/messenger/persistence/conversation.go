package persistence

import (
	"context"
	"database/sql"
	"demo/internal/errors"
	"demo/internal/messenger/domain"
	"encoding/base64"
	"encoding/json"
	"github.com/gofrs/uuid"
	"time"
)

var _ domain.ConversationRepository = &ConversationRepository{}

type ConversationRepository struct {
	queries *Queries
}

func NewConversationRepository(queries *Queries) *ConversationRepository {
	return &ConversationRepository{queries: queries}
}

func (c *ConversationRepository) Get(ctx context.Context, id uuid.UUID) (*domain.Conversation, error) {
	conversation, err := c.queries.GetConversation(ctx, id)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, domain.ErrConversationNotFound{ConversationId: id}
	}
	if err != nil {
		return nil, err
	}

	return &domain.Conversation{
		Id:         conversation.ID,
		SenderId:   conversation.SenderID,
		ReceiverId: conversation.ReceiverID,
		UpdatedAt:  conversation.UpdatedAt,
		CreatedAt:  conversation.CreatedAt,
	}, nil
}

func (c *ConversationRepository) Paginate(
	ctx context.Context,
	viewer uuid.UUID,
	first int,
	after sql.NullString,
) (*domain.Page[*domain.Conversation], error) {
	params := PaginateMyConversationsParams{
		ViewerID: viewer,
		First:    int32(first) + 1,
	}

	if after.Valid {
		cursor, err := c.decodeCursor(after.String)
		if err != nil {
			return nil, err
		}

		if rawUpdatedAt, ok := cursor["updatedAt"]; ok {
			params.AfterUpdatedAt = sql.NullTime{
				Time:  time.UnixMicro(int64(rawUpdatedAt.(float64))),
				Valid: true,
			}

			if rawId, ok := cursor["id"]; ok {
				id, err := uuid.FromString(rawId.(string))
				if err != nil {
					return nil, err
				}

				params.AfterID = uuid.NullUUID{
					UUID:  id,
					Valid: true,
				}
			}
		}
	}

	conversations, err := c.queries.PaginateMyConversations(ctx, params)
	if err != nil {
		return nil, err
	}

	edges := make([]domain.Edge[*domain.Conversation], 0, len(conversations))
	for i, conversation := range conversations {
		if i == first {
			break
		}

		cursor, err := c.encodeCursor(map[string]any{
			"updatedAt": conversation.UpdatedAt.UnixMicro(),
			"id":        conversation.ID.String(),
		})
		if err != nil {
			return nil, err
		}

		edges = append(edges, domain.Edge[*domain.Conversation]{
			Cursor: cursor,
			Node:   c.mapToDomain(conversation),
		})
	}

	var startCursor, endCursor *string
	if len(edges) > 0 {
		if len(edges) > 0 {
			startCursor = &edges[0].Cursor
			endCursor = &edges[len(edges)-1].Cursor
		}
	}

	return &domain.Page[*domain.Conversation]{
		Edges:           edges,
		HasNextPage:     len(conversations) > len(edges),
		HasPreviousPage: after.Valid,
		StartCursor:     startCursor,
		EndCursor:       endCursor,
	}, nil
}

func (c *ConversationRepository) GetOrCreate(ctx context.Context, sender uuid.UUID, receiver uuid.UUID) (*domain.Conversation, error) {
	conversation, err := c.queries.CreateOrGetConversation(ctx, CreateOrGetConversationParams{
		ID:         uuid.Must(uuid.NewV6()),
		SenderID:   sender,
		ReceiverID: receiver,
		UpdatedAt:  time.Now(),
		CreatedAt:  time.Now(),
	})

	if err != nil {
		return nil, err
	}

	return &domain.Conversation{
		Id:         conversation.ID,
		SenderId:   conversation.SenderID,
		ReceiverId: conversation.ReceiverID,
		UpdatedAt:  conversation.UpdatedAt,
		CreatedAt:  conversation.CreatedAt,
	}, nil
}

func (c *ConversationRepository) encodeCursor(cursor map[string]any) (string, error) {
	rawJson, err := json.Marshal(cursor)
	if err != nil {
		return "", err
	}
	return base64.StdEncoding.EncodeToString(rawJson), nil
}

func (c *ConversationRepository) decodeCursor(rawCursor string) (map[string]any, error) {
	rawJson, err := base64.StdEncoding.DecodeString(rawCursor)
	if err != nil {
		return nil, err
	}

	cursor := make(map[string]any)
	err = json.Unmarshal(rawJson, &cursor)
	if err != nil {
		return nil, err
	}

	return cursor, nil
}

func (c *ConversationRepository) mapToDomain(conversation MessengerConversation) *domain.Conversation {
	return &domain.Conversation{
		Id:         conversation.ID,
		SenderId:   conversation.SenderID,
		ReceiverId: conversation.ReceiverID,
		UpdatedAt:  conversation.UpdatedAt,
		CreatedAt:  conversation.CreatedAt,
	}
}
