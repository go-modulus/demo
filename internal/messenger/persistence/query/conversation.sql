-- name: GetConversation :one
select *
from "messenger"."conversation"
where id = $1
limit 1;

-- name: CreateOrGetConversation :one
with new_conversation as (
    insert into "messenger"."conversation" (id, sender_id, receiver_id, updated_at, created_at)
        select @id, @sender_id, @receiver_id, @updated_at, @created_at
        where not exists(
                select 1
                from "messenger"."conversation"
                where (sender_id = @sender_id and receiver_id = @receiver_id)
                   or (receiver_id = @sender_id and sender_id = @receiver_id)
            )
        on conflict do nothing
        returning *) (select *
                      from "messenger"."conversation"
                      where (sender_id = @sender_id and receiver_id = @receiver_id)
                         or (receiver_id = @sender_id and sender_id = @receiver_id)

                      union all

                      select *
                      from new_conversation) limit 1;

-- name: PaginateMyConversations :many
select *
from "messenger"."conversation"
where (sender_id = @viewer_id or receiver_id = @viewer_id)
  and (
        sqlc.narg(after_updated_at)::timestamp is null
        or updated_at < @after_updated_at
        or (sqlc.narg(after_id)::uuid is null or (updated_at = @after_updated_at and id > @after_id))
    )
order by updated_at desc, id
limit @first;