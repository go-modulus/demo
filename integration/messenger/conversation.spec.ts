import {persons} from "../utils";
import {getOrCreatePool, performScheduledDeletions, scheduleDeletion, sql} from "../db";
import {NotFoundError} from "slonik";
import {
    createConversation,
    revertCreateConversation,
    sendCreateOneToOneConversationMutation
} from "./conversation/create";
import {sendMyConversationsQuery} from "./conversation/my";

const clearDb = async () => {
    const pool = await getOrCreatePool();

    await pool.connect(async (connection) => {
        const personIds = Object.values(persons)

        try {
            const results = await connection.many(
                sql.typeAlias('id')`
                    SELECT id
                    FROM messenger.conversation
                    WHERE sender_id IN (${sql.join(personIds, sql.fragment`, `)})
                       OR receiver_id IN (${sql.join(personIds, sql.fragment`, `)})
                `
            )
            const conversationIds = results.map(({id}) => id)

            await connection.query(
                sql.typeAlias('void')`
                    DELETE
                    FROM messenger.conversation
                    WHERE id IN (${sql.join(conversationIds, sql.fragment`, `)})
                `
            )
        } catch (e) {
            if (!(e instanceof NotFoundError)) {
                throw e;
            }
        }
    })
}

beforeAll(clearDb)
afterEach(performScheduledDeletions)
afterAll(async () => {
    const pool = await getOrCreatePool();
    if (pool.getPoolState().ended) {
        return;
    }
    await pool.end();
})

test("create conversation", async () => {
    const res = await sendCreateOneToOneConversationMutation(persons.alice, persons.iggy);

    expect(res.error).toBeFalsy();
    expect(res.body).toMatchSnapshot({
        data: {
            createOneToOneConversation: {
                id: expect.any(String),
                createdAt: expect.any(Number),
            }
        }
    });

    const conversationId = res.body.data?.createOneToOneConversation.id;
    scheduleDeletion(() => revertCreateConversation(conversationId))
});

test("list conversations", async () => {
    const conversationIds = await Promise.all([
        createConversation(persons.alice, persons.bob),
        createConversation(persons.alice, persons.dawid),
        createConversation(persons.iggy, persons.alice),
    ])

    const res = await sendMyConversationsQuery(persons.alice);

    expect(res.error).toBeFalsy();
    expect(res.body).toMatchSnapshot({
        data: {
            myConversations: {
                edges: conversationIds.map(() => ({
                    cursor: expect.any(String),
                    node: {
                        id: expect.any(String),
                        createdAt: expect.any(Number),
                    },
                })),
                pageInfo: {
                    startCursor: expect.any(String),
                    endCursor: expect.any(String),
                },
            },
        },
    });

    const conversationIdsInResponse = res.body.data.myConversations.edges.map(
        (edge: any) => edge.node.id
    );

    expect(conversationIdsInResponse).toIncludeAllMembers(conversationIds);
});