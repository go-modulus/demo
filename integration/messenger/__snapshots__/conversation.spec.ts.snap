// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create conversation 1`] = `
{
  "data": {
    "createOneToOneConversation": {
      "createdAt": Any<Number>,
      "draft": null,
      "id": Any<String>,
      "lastMessage": null,
    },
  },
}
`;

exports[`list conversations 1`] = `
{
  "data": {
    "myConversations": {
      "edges": [
        {
          "cursor": Any<String>,
          "node": {
            "__typename": "OneToOneConversation",
            "createdAt": Any<Number>,
            "draft": null,
            "id": Any<String>,
            "lastMessage": null,
          },
        },
        {
          "cursor": Any<String>,
          "node": {
            "__typename": "OneToOneConversation",
            "createdAt": Any<Number>,
            "draft": null,
            "id": Any<String>,
            "lastMessage": null,
          },
        },
        {
          "cursor": Any<String>,
          "node": {
            "__typename": "OneToOneConversation",
            "createdAt": Any<Number>,
            "draft": null,
            "id": Any<String>,
            "lastMessage": null,
          },
        },
      ],
      "pageInfo": {
        "endCursor": Any<String>,
        "hasNextPage": false,
        "hasPreviousPage": false,
        "startCursor": Any<String>,
      },
    },
  },
}
`;
